sdk/core/Azure.Core/src/Pipeline/AzureOperationScope.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+
+namespace Azure.Core.Pipeline
+{
+    public struct ClientDiagnosticScope: IDisposable
+    {
+        private Activity _activity;
+
+        private readonly string _name;
+
+        private readonly DiagnosticListener _source;
+
+        internal ClientDiagnosticScope(string name, DiagnosticListener source)
+        {
+            _name = name;
+            _source = source;
+            _activity = _source.IsEnabled() ? new Activity(_name) : null;
+        }
+
+        public bool IsEnabled => _activity != null;
+
+        public void AddAttribute(string name, string value)
+        {
+            _activity?.AddTag(name, value);
+        }
+
+        public void AddAttribute<T>(string name, T value, Func<T, string> format)
```

Comment:
Why not just a `Func<string>` that can do whatever?

Reply:
Avoids closure allocation. By avoiding capturing locals closure would be allocated only once and cached by the compiler.

Reply:
Do we expect customers to use this API for adding their own scopes around sets of operations?  It might be worth having easier to use/understand overloads for them to use (where the allocations matter less because they'll definitely be tracing if they add the calls).  I'm fine waiting on that for the time being.

Should we also add an overload that takes an object and calls ToString() to avoid any allocations for the most common scenario?  I imagine most of the things we'll add as attributes are either primitives or have their own ToString implementation.

Reply:
> Should we also add an overload that takes an object and calls ToString() to avoid any allocations for the most common scenario? I imagine most of the things we'll add as attributes are either primitives or have their own ToString implementation.

Great point.

> Do we expect customers to use this API for adding their own scopes around sets of operations? It might be worth having easier to use/understand overloads for them to use (where the allocations matter less because they'll definitely be tracing if they add the calls). I'm fine waiting on that for the time being.

No, customers should be using Activity directly.